vllm/attention/backends/xformers_cla.py

```diff
@@ -0,0 +1,927 @@
+                        # kv_cache is not created yet during profiling, so we need to create key and value.
+                        key = torch.zeros(query.size(0), self.num_kv_heads, self.head_size, dtype=query.dtype, device=query.device)
+                        value = torch.zeros(query.size(0), self.num_kv_heads, self.head_size, dtype=query.dtype, device=query.device)
+                    prefill_meta.shared_self_attention_types.append(SharedSelfAttentionType.PREFILL_PROFILING.name)
+
+                out = self._run_memory_efficient_xformers_forward(query, key, value, prefill_meta, attn_type=attn_type)
+                assert out.shape == output[:num_prefill_tokens].shape
+                output[:num_prefill_tokens] = out
+            else:
+                # TODO: Test this KV Cache Sharing. 
+                # print("Prefilling with KV cache (prefix-cache enabled)")
+                # print(prefill_meta.block_tables.shape, kv_cache.shape if kv_cache is not None else 'None')
+                # print(prefill_meta.block_tables)
+                assert prefill_meta.query_start_loc is not None
+                assert prefill_meta.max_query_len is not None
+                
+                if kv_cache.numel() > 0 and not self.compute_new_kv:
+                    # Split kv_cache into key_cache and value_cache.
+                    key_cache, value_cache = PagedAttention.split_kv_cache(kv_cache, self.num_kv_heads, self.head_size)
+                    block_size = value_cache.size(-1)
+                    
+                    # Extract KV from cache using slot_mapping.
+                    k_reshaped, v_reshaped = convert_cache_to_original_shape(key_cache, value_cache, self.num_kv_heads, self.head_size)
+                    slot_mapping = prefill_meta.slot_mapping
+                    block_idxs, token_idxs = slot_mapping // block_size, slot_mapping % block_size
+                    key, value = k_reshaped[block_idxs, token_idxs], v_reshaped[block_idxs, token_idxs]
+                    if self.kv_cache_dtype in ("fp8", "fp8_e4m3"):
+                        key = (key.view(torch.float8_e4m3fn).to(torch.float32) * k_scale).to(query.dtype)
+                        value = (value.view(torch.float8_e4m3fn).to(torch.float32) * v_scale).to(query.dtype)
+                    elif self.kv_cache_dtype == "fp8_e5m2":
+                        key = (key.view(torch.float8_e5m2).to(torch.float32) * k_scale).to(query.dtype)
+                        value = (value.view(torch.float8_e5m2).to(torch.float32) * v_scale).to(query.dtype)
+
+                # prefix-enabled attention
+                # TODO(Hai) this triton kernel has regression issue (broke) to
+                # deal with different data types between KV and FP8 KV cache,
+                # to be addressed separately.
+                # print(f"query.shape: {query.shape}, key.shape: {key.shape}, value.shape: {value.shape}")
+                out = PagedAttention.forward_prefix(
```

Comment:
This isn't a suggestion for PR but more of a general question.

Do you know why you need to pass in both `key`, `value`, as well as `key_cache`, `value_cache` ?

It seems like one or the other should be necessary.

If we didn't need `key` and `value` we wouldn't need to extract it and reshape it

Reply:
This is something we need to figure out eventually but my guess is -> when prefix caching is enabled and if there is a partial match between the prompt and the cache. For example, a prompt with 150 tokens might have its first 100 tokens cached and the remaining 50 tokens might be new/unique. 

Looking at the prefill triton kernel there are 2 sections 1) [compute query against context (no causal mask here)](https://github.com/AnswerDotAI/vllm/blob/e4c848691c31f46e5a3635470166bce3137e17ff/vllm/attention/ops/prefix_prefill.py#L101-L126) and 2) [compute query against itself (with causal mask)](https://github.com/AnswerDotAI/vllm/blob/e4c848691c31f46e5a3635470166bce3137e17ff/vllm/attention/ops/prefix_prefill.py#L196-L208)

Reply:
Attention layer forward pass is called even when there is a prefix cache hit, which means q,k,v are recomputed which deceives the whole point of avoiding recomputations. Maybe we should ask about this to have a better understanding.

Reply:
Thanks - making a note that we discussed this on discord and have a shared understanding :)